contracts/infrastructure/helper/MultiCallHelper.sol

```diff
@@ -0,0 +1,148 @@
+    bytes4 private constant ERC721_SAFE_TRANSFER_FROM_BYTES = bytes4(keccak256("safeTransferFrom(address,address,uint256,bytes)"));
+    bytes4 private constant ERC721_SET_APPROVAL_FOR_ALL = bytes4(keccak256("setApprovalForAll(address,bool)"));
+    bytes4 private constant ERC1155_SAFE_TRANSFER_FROM = bytes4(keccak256("safeTransferFrom(address,address,uint256,uint256,bytes)"));
+
+    struct Call {
+        address to;
+        uint256 value;
+        bytes data;
+    }
+
+    // The trusted contacts storage
+    ITransferStorage internal immutable userWhitelist;
+    // The dapp registry contract
+    IDappRegistry internal immutable dappRegistry;
+
+    constructor(ITransferStorage _userWhitelist, IDappRegistry _dappRegistry) public {
+        userWhitelist = _userWhitelist;
+        dappRegistry = _dappRegistry;
+    }
+
+    /**
+     * @notice Checks if a sequence of transactions is authorised to be executed by a wallet.
+     * The method returns false if any of the inner transaction is not to a trusted contact or an authorised dapp.
+     * @param _wallet The target wallet.
+     * @param _transactions The sequence of transactions.
+     */
+    function isMultiCallAuthorised(address _wallet, Call[] calldata _transactions) external view returns (bool) {
+        for(uint i = 0; i < _transactions.length; i++) {
+            address spender = recoverSpender(_wallet, _transactions[i]);
+            if (!isWhitelisted(_wallet, spender) && isAuthorised(_wallet, spender, _transactions[i].to, _transactions[i].data) == MAX_UINT) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    /**
+     * @notice Checks if each of the transaction of a sequence of transactions is authorised to be executed by a wallet.
+     * For each transaction of the sequence it returns an Id where:
+     *     - Id is in [0,255]: the transaction is to an address authorised in registry Id of the DappRegistry
+     *     - Id = 256: the transaction is to an address authorised in the trusted contacts of the wallet
+     *     - Id = MAX_UINT: the transaction is not authorised
+     * @param _wallet The target wallet.
+     * @param _transactions The sequence of transactions.
+     */
+    function multiCallAuthorisation(address _wallet, Call[] calldata _transactions) external view returns (uint256[] memory registryIds) {
+        registryIds = new uint256[](_transactions.length);
+        for(uint i = 0; i < _transactions.length; i++) {
+            address spender = recoverSpender(_wallet, _transactions[i]);
+            if (isWhitelisted(_wallet, spender)) {
+                registryIds[i] = 256;
+            } else {
+                registryIds[i] = isAuthorised(_wallet, spender, _transactions[i].to, _transactions[i].data);
+            }
+        }
+    }
+
+    function recoverSpender(address _wallet, Call calldata _transaction) internal pure returns (address) {
```

Comment:
Can abstract that in `Utils` library. It'll save some bytecode off the module as well that way.